### PR TITLE
Deprecate Python 3.6

### PR DESCRIPTION
**Pull request by tony**
Resolves: https://github.com/python-poetry/poetry/issues/4997
See also: https://github.com/python-poetry/poetry/pull/5055, https://github.com/python-poetry/install.python-poetry.org/pull/11

As I'm getting acclimated with the code base, this may have some back and forth to it

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

Later?
- [ ] tomlkit: Potentially update port of tomlkit to [0.8.0](https://github.com/sdispater/tomlkit/releases/tag/0.8.0), diff since: [0.7.2](https://github.com/sdispater/tomlkit/compare/0.7.2...0.8.0), [0.7.2](https://github.com/sdispater/tomlkit/compare/0.7.2...0.8.0), [0.6.0](https://github.com/sdispater/tomlkit/compare/0.6.0...0.8.0), [0.5.11](https://github.com/sdispater/tomlkit/compare/0.5.11...0.8.0)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -33,23 +33,22 @@ exclude = [
...
-python = "^3.6"
+python = "^3.7"
 
 # required for compatibility
 importlib-metadata = {version = ">=1.7.0", python = "<3.8"}
-dataclasses = {version = ">=0.8", python = "~3.6"}
 
 [tool.poetry.dev-dependencies]
-pre-commit = {version = "^2.15.0", python = "^3.6.1"}
+pre-commit = {version = "^2.15.0", python = "^3.7.0"}
```
Comment (branchvincent) [inline pyproject.toml:42]: similar change over here:
```suggestion
pre-commit = "^2.15.0"
```

File: pyproject.toml [hunk 2]
```diff
@@ -33,23 +33,22 @@ exclude = [
...
 pytest = "^6.2"
 pytest-cov = "^2.8"
 pytest-mock = "^3.5"
 tox = "^3.0"
 vendoring = {version = "^0.3", python = "^3.8"}
 pep517 = "^0.11.0"
-black = {version = "^21.9b0", markers = "python_version >= '3.6.2' and python_version < '4.0' and implementation_name != 'pypy'"}
-isort = {version = "^5.9.3", python = "^3.6.1"}
+black = {version = "^21.9b0", markers = "python_version >= '3.7.0' and python_version < '4.0' and implementation_name != 'pypy'"}
+isort = {version = "^5.9.3"}
```
Comment (branchvincent) [inline pyproject.toml:51]: just to be consistent:
```suggestion
isort = "^5.9.3"
```

File: tests/testutils.py [hunk 3]
```diff
@@ -57,9 +57,7 @@ def subprocess_run(*args: str, **kwargs: Any) -> subprocess.CompletedProcess:
     Helper method to run a subprocess. Asserts for success.
     """
     compat_kwargs = {"text" if PY37 else "universal_newlines": True}
-    result = subprocess.run(
-        args, **compat_kwargs, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs
-    )
+    result = subprocess.run(args, **compat_kwargs, capture_output=True, **kwargs)
```
Comment (branchvincent) [inline tests/testutils.py:60]: should be able to drop `compat_kwargs`:
```suggestion
    result = subprocess.run(args, text=True, capture_output=True, **kwargs)
```

File: src/poetry/core/utils/_compat.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 from typing import List
 
 
-PY36 = sys.version_info >= (3, 6)
 PY37 = sys.version_info >= (3, 7)
```
Comment (branchvincent) [inline src/poetry/core/utils/_compat.py:6]: we should be able to drop `PY37` and any references (I see a couple in the tests)

**Review by branchvincent — CHANGES_REQUESTED** [review]
Thanks @tony, couple suggestions

File: tests/testutils.py (hunk 3 above)
Comment (tony) [reply tests/testutils.py:60]: Good catch

File: pyproject.toml (hunk 2 above)
Comment (tony) [reply pyproject.toml:51]: Another good catch

File: src/poetry/core/utils/_compat.py (hunk 4 above)
Comment (tony) [reply src/poetry/core/utils/_compat.py]: ec61d70

File: src/poetry/core/utils/_compat.py [hunk 5]
```diff
@@ -3,9 +3,6 @@
 from typing import List
 
 
-PY36 = sys.version_info >= (3, 6)
-PY37 = sys.version_info >= (3, 7)
-
 WINDOWS = sys.platform == "win32"
```
Comment (danieleades) [inline src/poetry/core/utils/_compat.py:8]: is the following guard still required?

```python
try:
    FileNotFoundError
except NameError:
    FileNotFoundError = IOError  # noqa
```
?

**Review by danieleades — CHANGES_REQUESTED** [review]

File: src/poetry/core/utils/_compat.py (hunk 5 above)
Comment (tony) [reply src/poetry/core/utils/_compat.py:8]: Good catch, we're not using it

**Review by danieleades — APPROVED** [review]

**Review by danieleades — DISMISSED** [review]

**Comment by tony** [comment]
@neersighted can this have tests ran on them?  

Is it possible to allow tests to always run for me?

**Comment by tony** [comment]
in re:

> @tony tony dismissed danieleades’s stale review via 4bb19c9 1 minute ago

@danieleades Just a rebase against master

**Comment by neersighted** [comment]
> @neersighted can this have tests ran on them?  
> 
> Is it possible to allow tests to always run for me?

There is no manual whitelist. Having a merged PR will automatically run CI for you in the future.

**Review by danieleades — APPROVED** [review]

**Review by finswimmer — APPROVED** [review]

**Pull request merged**